subprojects/dependency-management/src/main/java/org/gradle/internal/locking/DefaultDependencyLockingProvider.java

```diff
@@ -139,19 +172,31 @@ private ModuleComponentIdentifier parseLockNotation(String configurationName, St
     public void persistResolvedDependencies(String configurationName, Set<ModuleComponentIdentifier> resolvedModules, Set<ModuleComponentIdentifier> changingResolvedModules) {
         if (writeLocks) {
             List<String> modulesOrdered = getModulesOrdered(resolvedModules);
-            lockFileReaderWriter.writeLockFile(configurationName, modulesOrdered);
             if (!changingResolvedModules.isEmpty()) {
                 LOGGER.warn("Dependency lock state for configuration '{}' contains changing modules: {}. This means that dependencies content may still change over time. See {} for details.",
                     context.identityPath(configurationName), getModulesOrdered(changingResolvedModules), DOC_REG.getDocumentationFor("dependency_locking"));
             }
-            if (LOGGER.isDebugEnabled()) {
-                LOGGER.debug("Persisted dependency lock state for configuration '{}', state is: {}", context.identityPath(configurationName), modulesOrdered);
+            if (uniqueLockStateEnabled) {
+                allLockState.put(configurationName, modulesOrdered);
             } else {
-                LOGGER.lifecycle("Persisted dependency lock state for configuration '{}'", context.identityPath(configurationName));
+                lockFileReaderWriter.writeLockFile(configurationName, modulesOrdered);
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug("Persisted dependency lock state for configuration '{}', state is: {}", context.identityPath(configurationName), modulesOrdered);
+                } else {
+                    LOGGER.lifecycle("Persisted dependency lock state for configuration '{}'", context.identityPath(configurationName));
+                }
             }
         }
     }
 
+    @Override
+    public void buildFinished() {
+        if (uniqueLockStateEnabled && uniqueLockStateLoaded && lockFileReaderWriter.canWrite()) {
+            lockFileReaderWriter.writeUniqueLockfile(allLockState);
+            LOGGER.lifecycle("Persisted dependency lock state for project '{}' (buildscript: {})", context.getProjectPath(), context.isScript());
```

Comment:
:o: I stumbed upon this message when locking a test project when reviewing and I found that this `buildscript: false` is a bit confusing. Can you make it more explicit, like "Persisted build script lock state"?